create-strapi-app --quickstart fails with unable to resolve dependency tree

<!--
Hello 👋 Thank you for submitting an issue.

Before you start, please make sure your issue is understandable and reproducible.
To make your issue readable make sure you use valid Markdown syntax.

https://guides.github.com/features/mastering-markdown/

Please ensure you have also read and understand the contributing guide.

https://github.com/strapi/strapi/blob/main/CONTRIBUTING.md#reporting-an-issue
-->

## Bug report

### Required System information

<!-- Please ensure you are using the Node LTS version (v18 or v20) -->
<!-- Strapi v3 is no longer supported, please update to Strapi v4 -->
<!-- If you are reporting a frontend bug please provide error logs after setting STRAPI_ENFORCE_SOURCEMAPS=true in your .env -->
<!-- This environment variable makes frontend errors easier to read and trace -->

- Node.js version: `v20.9.0`
- NPM version: `10.2.2`
- Strapi version: `4.15.1`
- Database: sqlite
- Operating system: `macOS 14.1`
- Is your project Javascript or Typescript: JavaScript

<!-- Did you know you can run `yarn strapi report` or `npm run strapi report` to get this information easier? -->

### Describe the bug

Creating a new Strapi project fails due to dependency issues.

### Steps to reproduce the behavior

1. Run `npx create-strapi-app@latest my-project --quickstart` as per [Quick Start Guide ](https://docs.strapi.io/dev-docs/quick-start)
2. There is no step 2. Step 1 fails with 
```
% npx create-strapi-app@latest my-project --quickstart
Creating a quickstart project.
Creating a new Strapi application at /Users/mirko/Projects/my-project.
Creating files.
Error while installing dependencies:
npm WARN config production Use `--omit=dev` instead.
npm WARN config optional Use `--omit=optional` to exclude optional dependencies, or
npm WARN config `--include=optional` to include them.
npm WARN config 
npm WARN config     Default value does install optional deps unless otherwise omitted.
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! While resolving: my-project@0.1.0
npm ERR! Found: styled-components@5.3.3
npm ERR! node_modules/styled-components
npm ERR!   styled-components@"5.3.3" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer styled-components@"^5.3.6" from @strapi/plugin-cloud@4.15.1
npm ERR! node_modules/@strapi/plugin-cloud
npm ERR!   @strapi/plugin-cloud@"4.15.1" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
```

### Expected behavior

`npm install` should complete without errors.

### Screenshots

If applicable, add screenshots to help explain your problem.

### Code snippets

If applicable, add code samples to help explain your problem.

### Additional context

Add any other context about the problem here.

getting a similar error right after 4.15.1 dropped, not an issue for me on 4.15.0. can't run an install and build or create a new project anymore without getting this error and I'm not able to force `~14.5.0` anymore for some reason in my package.json. I've added the suggested react and styled components dependencies too. please revert `14.5.1` or correct the dependency issue, all of my strapi projects are unusable currently and this has happened many times where a non-breaking change version introduces breaking changes.
```
Building admin panel[ERROR] Module not found: Error: Package path ./admin/src/content-manager/utils is not exported from package ~/Documents/work-files/strapi/backend/node_modules/@strapi/admin (see exports field in ~/Documents/work-files/strapi/backend/node_modules/@strapi/admin/package.json)
```

The generated project has `"styled-components": "5.3.3"` in its `dependencies`, but also `"@strapi/plugin-cloud": "4.15.1"` which has `'styled-components': '^5.3.6'` in its `peerDependencies`. 

However, setting `styled-components` to `"5.3.6"` in `package.json` then fails because `@strapi/plugin-users-permissions@4.15.1` expects `styled-components@"5.3.3"`

```
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! While resolving: my-project@0.1.0
npm ERR! Found: styled-components@5.3.6
npm ERR! node_modules/styled-components
npm ERR!   styled-components@"5.3.6" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer styled-components@"5.3.3" from @strapi/plugin-users-permissions@4.15.1
npm ERR! node_modules/@strapi/plugin-users-permissions
npm ERR!   @strapi/plugin-users-permissions@"4.15.1" from the root project
npm ERR! 
```

So `plugin-cloud` and `plugin-users-permissions` are incompatible with one another, even though they're both `4.15.1`.

A workaround is to remove the `"@strapi/plugin-cloud"` dependency altogether.

It's not really easy to test, because a `0.0.0-*` version is not compatible with `^4.4.0`.

```
% npx create-strapi-app@0.0.0-experimental.1cec6226fe6c8ca4b8404f1ebac8076d5ee5b336 my-project --quickstart
Creating a quickstart project.
Creating a new Strapi application at /Users/mirko/Projects/my-project.
Creating files.
Error while installing dependencies:
npm WARN config production Use `--omit=dev` instead.
npm WARN config optional Use `--omit=optional` to exclude optional dependencies, or
npm WARN config `--include=optional` to include them.
npm WARN config 
npm WARN config     Default value does install optional deps unless otherwise omitted.
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! While resolving: my-project@0.1.0
npm ERR! Found: @strapi/strapi@0.0.0-experimental.1cec6226fe6c8ca4b8404f1ebac8076d5ee5b336
npm ERR! node_modules/@strapi/strapi
npm ERR!   @strapi/strapi@"0.0.0-experimental.1cec6226fe6c8ca4b8404f1ebac8076d5ee5b336" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer @strapi/strapi@"^4.4.0" from @strapi/plugin-cloud@0.0.0-experimental.1cec6226fe6c8ca4b8404f1ebac8076d5ee5b336
npm ERR! node_modules/@strapi/plugin-cloud
npm ERR!   @strapi/plugin-cloud@"0.0.0-experimental.1cec6226fe6c8ca4b8404f1ebac8076d5ee5b336" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
```

--- a/packages/core/admin/_internal/node/core/dependencies.ts
+++ b/packages/core/admin/_internal/node/core/dependencies.ts
@@ -15,8 +15,8 @@ import { getPackageManager } from './managers';
 const PEER_DEPS = {
   react: '^18.0.0',
   'react-dom': '^18.0.0',
-  'react-router-dom': '^5.0.0',
-  'styled-components': '^5.0.0',
+  'react-router-dom': '^5.2.0',
+  'styled-components': '^5.2.1',
 };
 
 interface CheckRequiredDependenciesResult {

--- a/packages/core/admin/_internal/node/core/plugins.ts
+++ b/packages/core/admin/_internal/node/core/plugins.ts
@@ -5,13 +5,6 @@ import { getModule, PackageJson } from './dependencies';
 import { loadFile } from './files';
 import { BuildContext, CreateBuildContextArgs } from '../createBuildContext';
 
-const CORE_PLUGINS = [
-  '@strapi/plugin-content-manager',
-  '@strapi/plugin-content-type-builder',
-  '@strapi/plugin-email',
-  '@strapi/plugin-upload',
-];
-
 interface PluginMeta {
   name: string;
   pathToPlugin: string;
@@ -45,24 +38,6 @@ export const getEnabledPlugins = async ({
   logger,
 }: Pick<BuildContext, 'cwd' | 'logger' | 'strapi'>) => {
   const plugins: Record<string, PluginMeta> = {};
-
-  logger.debug('Core plugins', os.EOL, CORE_PLUGINS);
-
-  for (const plugin of CORE_PLUGINS) {
-    const pkg = await getModule(plugin, cwd);
-
-    if (pkg && validatePackageIsPlugin(pkg)) {
-      /**
-       * We know there's a name because these are our packages.
-       */
-      const name = (pkg.strapi.name || pkg.name)!;
-
-      plugins[name] = {
-        name,
-        pathToPlugin: plugin,
-      };
-    }
-  }
 
   /**
    * This is the list of dependencies that are installed in the user's project.

--- a/packages/core/admin/_internal/node/staticFiles.ts
+++ b/packages/core/admin/_internal/node/staticFiles.ts
@@ -26,7 +26,7 @@ const getEntryModule = ({ plugins }: EntryModuleArgs): string => {
          * Any modifications made will be discarded.
          */
         ${pluginsImport}
-        import { renderAdmin } from "@strapi/admin/strapi-admin"
+        import { renderAdmin } from "@strapi/strapi/admin"
         
         renderAdmin(
           document.getElementById("strapi"),

--- a/packages/core/admin/package.json
+++ b/packages/core/admin/package.json
@@ -96,6 +96,7 @@
     "css-loader": "^6.8.1",
     "date-fns": "2.30.0",
     "dotenv": "14.2.0",
+    "esbuild": "0.19.2",
     "esbuild-loader": "^2.21.0",
     "esbuild-register": "3.5.0",
     "execa": "5.1.1",
@@ -185,7 +186,12 @@
     "vite": "4.4.9"
   },
   "peerDependencies": {
-    "@strapi/strapi": "^4.3.4"
+    "@strapi/data-transfer": "4.15.0",
+    "@strapi/strapi": "^4.3.4",
+    "react": "^17.0.0 || ^18.0.0",
+    "react-dom": "^17.0.0 || ^18.0.0",
+    "react-router-dom": "^5.2.0",
+    "styled-components": "^5.2.1"
   },
   "engines": {
     "node": ">=18.0.0 <=20.x.x",

--- a/packages/core/content-type-builder/package.json
+++ b/packages/core/content-type-builder/package.json
@@ -58,8 +58,8 @@
   "peerDependencies": {
     "react": "^17.0.0 || ^18.0.0",
     "react-dom": "^17.0.0 || ^18.0.0",
-    "react-router-dom": "5.3.4",
-    "styled-components": "5.3.3"
+    "react-router-dom": "^5.2.0",
+    "styled-components": "^5.2.1"
   },
   "engines": {
     "node": ">=18.0.0 <=20.x.x",

--- a/packages/core/data-transfer/src/commands/data-transfer.ts
+++ b/packages/core/data-transfer/src/commands/data-transfer.ts
@@ -5,7 +5,7 @@ import { configs, createLogger } from '@strapi/logger';
 import strapiFactory from '@strapi/strapi';
 import ora from 'ora';
 import { merge } from 'lodash/fp';
-import type { LoadedStrapi } from '@strapi/types';
+import type { LoadedStrapi, Strapi } from '@strapi/types';
 
 import { readableBytes, exitWith } from './helpers';
 import { getParseListWithChoices, parseInteger, confirmMessage } from './commander';
@@ -148,7 +148,9 @@ const setSignalHandler = async (
   });
 };
 
-const createStrapiInstance = async (opts: { logLevel?: string } = {}) => {
+const createStrapiInstance = async (
+  opts: { logLevel?: string } = {}
+): Promise<Strapi & Required<Strapi>> => {
   try {
     const appContext = await strapiFactory.compile();
     const app = strapiFactory({ ...opts, ...appContext });

--- a/packages/core/email/package.json
+++ b/packages/core/email/package.json
@@ -81,8 +81,8 @@
     "koa": "2.13.4",
     "react": "^17.0.0 || ^18.0.0",
     "react-dom": "^17.0.0 || ^18.0.0",
-    "react-router-dom": "5.3.4",
-    "styled-components": "5.3.3"
+    "react-router-dom": "^5.2.0",
+    "styled-components": "^5.2.1"
   },
   "engines": {
     "node": ">=18.0.0 <=20.x.x",

--- a/packages/core/helper-plugin/package.json
+++ b/packages/core/helper-plugin/package.json
@@ -100,8 +100,8 @@
     "@strapi/icons": "1.13.0",
     "react": "^17.0.0 || ^18.0.0",
     "react-dom": "^17.0.0 || ^18.0.0",
-    "react-router-dom": "^5.3.4",
-    "styled-components": "^5.3.3"
+    "react-router-dom": "^5.2.0",
+    "styled-components": "^5.2.1"
   },
   "engines": {
     "node": ">=18.0.0 <=20.x.x",

--- a/packages/core/strapi/package.json
+++ b/packages/core/strapi/package.json
@@ -64,6 +64,30 @@
       "url": "https://strapi.io"
     }
   ],
+  "exports": {
+    ".": {
+      "types": "./dist/index.d.ts",
+      "source": "./src/index.ts",
+      "import": "./dist/index.mjs",
+      "require": "./dist/index.js",
+      "default": "./dist/index.js"
+    },
+    "./dist/utils/ee": {
+      "types": "./dist/utils/ee.d.ts",
+      "source": "./src/utils/ee.ts",
+      "import": "./dist/utils/ee.mjs",
+      "require": "./dist/utils/ee.js",
+      "default": "./dist/utils/ee.js"
+    },
+    "./admin": {
+      "types": "./dist/admin.d.ts",
+      "source": "./src/admin.ts",
+      "import": "./dist/admin.mjs",
+      "require": "./dist/admin.js",
+      "default": "./dist/admin.js"
+    },
+    "./package.json": "./package.json"
+  },
   "main": "./dist/index.js",
   "module": "./dist/index.mjs",
   "source": "./src/index.ts",

--- a/packages/core/strapi/packup.config.ts
+++ b/packages/core/strapi/packup.config.ts
@@ -5,12 +5,27 @@ import { builtinModules } from 'node:module';
 export default defineConfig({
   bundles: [
     {
+      source: './src/index.ts',
+      import: './dist/index.js',
+      require: './dist/index.js',
+      types: './dist/index.d.ts',
+      runtime: 'node',
+    },
+    {
       source: './src/cli.ts',
       require: './dist/cli.js',
       runtime: 'node',
     },
+    {
+      source: './src/admin.ts',
+      import: './dist/admin.js',
+      require: './dist/admin.js',
+      types: './dist/index.d.ts',
+      runtime: 'web',
+    },
   ],
+  exports: {},
+  dist: './dist',
   externals: [...builtinModules],
   preserveModules: true,
-  runtime: 'node',
 });

--- a/packages/core/strapi/src/admin.ts
+++ b/packages/core/strapi/src/admin.ts
@@ -1,0 +1,21 @@
+import { RenderAdminArgs, renderAdmin } from '@strapi/admin/strapi-admin';
+// @ts-expect-error – No types, yet.
+import contentTypeBuilder from '@strapi/plugin-content-type-builder/strapi-admin';
+import email from '@strapi/plugin-email/strapi-admin';
+// @ts-expect-error – No types, yet.
+import upload from '@strapi/plugin-upload/strapi-admin';
+
+const render = (mountNode: HTMLElement | null, { plugins }: RenderAdminArgs) => {
+  return renderAdmin(mountNode, {
+    plugins: {
+      'content-type-builder': contentTypeBuilder,
+      // @ts-expect-error – TODO: fix this
+      email,
+      upload,
+      ...plugins,
+    },
+  });
+};
+
+export { render as renderAdmin };
+export type { RenderAdminArgs };

--- a/packages/core/strapi/src/commands/actions/plugin/build-command/action.ts
+++ b/packages/core/strapi/src/commands/actions/plugin/build-command/action.ts
@@ -13,6 +13,10 @@ export default async ({ force, ...opts }: ActionOptions) => {
   const logger = createLogger({ debug: opts.debug, silent: opts.silent, timestamp: false });
   try {
     /**
+     * ALWAYS set production for using plugin build CLI.
+     */
+    process.env.NODE_ENV = 'production';
+    /**
      * Notify users this is an experimental command and get them to approve first
      * this can be opted out by setting the argument --yes
      */

--- a/packages/core/upload/package.json
+++ b/packages/core/upload/package.json
@@ -65,8 +65,8 @@
   "peerDependencies": {
     "react": "^17.0.0 || ^18.0.0",
     "react-dom": "^17.0.0 || ^18.0.0",
-    "react-router-dom": "5.3.4",
-    "styled-components": "5.3.3"
+    "react-router-dom": "^5.2.0",
+    "styled-components": "^5.2.1"
   },
   "engines": {
     "node": ">=18.0.0 <=20.x.x",

--- a/packages/generators/generators/src/templates/js/plugin-package.json.hbs
+++ b/packages/generators/generators/src/templates/js/plugin-package.json.hbs
@@ -23,8 +23,8 @@
   "peerDependencies": {
     "react": "^17.0.0 || ^18.0.0",
     "react-dom": "^17.0.0 || ^18.0.0",
-    "react-router-dom": "^5.3.4",
-    "styled-components": "^5.3.6"
+    "react-router-dom": "^5.2.0",
+    "styled-components": "^5.2.1"
   },
   "author": {
     "name": "A Strapi developer"

--- a/packages/generators/generators/src/templates/ts/plugin-package.json.hbs
+++ b/packages/generators/generators/src/templates/ts/plugin-package.json.hbs
@@ -28,8 +28,8 @@
   "peerDependencies": {
     "react": "^17.0.0 || ^18.0.0",
     "react-dom": "^17.0.0 || ^18.0.0",
-    "react-router-dom": "^5.3.4",
-    "styled-components": "^5.3.6"
+    "react-router-dom": "^5.2.0",
+    "styled-components": "^5.2.1"
   },
   "author": {
     "name": "A Strapi developer"

--- a/packages/plugins/cloud/package.json
+++ b/packages/plugins/cloud/package.json
@@ -53,8 +53,8 @@
     "eslint-config-custom": "4.15.1",
     "react": "^18.2.0",
     "react-dom": "^18.2.0",
-    "react-router-dom": "^5.3.4",
-    "styled-components": "^5.3.3",
+    "react-router-dom": "5.3.4",
+    "styled-components": "5.3.3",
     "tsconfig": "4.15.1",
     "typescript": "5.2.2"
   },
@@ -62,8 +62,8 @@
     "@strapi/strapi": "^4.4.0",
     "react": "^17.0.0 || ^18.0.0",
     "react-dom": "^17.0.0 || ^18.0.0",
-    "react-router-dom": "^5.3.4",
-    "styled-components": "^5.3.6"
+    "react-router-dom": "^5.2.0",
+    "styled-components": "^5.2.1"
   },
   "engines": {
     "node": ">=18.0.0 <=20.x.x",

--- a/packages/plugins/color-picker/package.json
+++ b/packages/plugins/color-picker/package.json
@@ -79,8 +79,8 @@
     "@strapi/strapi": "^4.4.0",
     "react": "^17.0.0 || ^18.0.0",
     "react-dom": "^17.0.0 || ^18.0.0",
-    "react-router-dom": "5.3.4",
-    "styled-components": "5.3.3"
+    "react-router-dom": "^5.2.0",
+    "styled-components": "^5.2.1"
   },
   "engines": {
     "node": ">=18.0.0 <=20.x.x",

--- a/packages/plugins/documentation/package.json
+++ b/packages/plugins/documentation/package.json
@@ -61,8 +61,8 @@
     "@strapi/strapi": "^4.0.0",
     "react": "^17.0.0 || ^18.0.0",
     "react-dom": "^17.0.0 || ^18.0.0",
-    "react-router-dom": "5.3.4",
-    "styled-components": "5.3.3"
+    "react-router-dom": "^5.2.0",
+    "styled-components": "^5.2.1"
   },
   "engines": {
     "node": ">=18.0.0 <=20.x.x",

--- a/packages/plugins/graphql/package.json
+++ b/packages/plugins/graphql/package.json
@@ -54,8 +54,8 @@
     "@strapi/strapi": "^4.0.0",
     "react": "^17.0.0 || ^18.0.0",
     "react-dom": "^17.0.0 || ^18.0.0",
-    "react-router-dom": "5.3.4",
-    "styled-components": "5.3.3"
+    "react-router-dom": "^5.2.0",
+    "styled-components": "^5.2.1"
   },
   "engines": {
     "node": ">=18.0.0 <=20.x.x",

--- a/packages/plugins/i18n/package.json
+++ b/packages/plugins/i18n/package.json
@@ -56,8 +56,8 @@
   "peerDependencies": {
     "react": "^17.0.0 || ^18.0.0",
     "react-dom": "^17.0.0 || ^18.0.0",
-    "react-router-dom": "5.3.4",
-    "styled-components": "5.3.3"
+    "react-router-dom": "^5.2.0",
+    "styled-components": "^5.2.1"
   },
   "engines": {
     "node": ">=18.0.0 <=20.x.x",

--- a/packages/plugins/sentry/package.json
+++ b/packages/plugins/sentry/package.json
@@ -41,8 +41,8 @@
     "@strapi/strapi": "^4.0.0",
     "react": "^17.0.0 || ^18.0.0",
     "react-dom": "^17.0.0 || ^18.0.0",
-    "react-router-dom": "5.3.4",
-    "styled-components": "5.3.3"
+    "react-router-dom": "^5.2.0",
+    "styled-components": "^5.2.1"
   },
   "engines": {
     "node": ">=18.0.0 <=20.x.x",

--- a/packages/plugins/users-permissions/package.json
+++ b/packages/plugins/users-permissions/package.json
@@ -63,8 +63,8 @@
   "peerDependencies": {
     "react": "^17.0.0 || ^18.0.0",
     "react-dom": "^17.0.0 || ^18.0.0",
-    "react-router-dom": "5.3.4",
-    "styled-components": "5.3.3"
+    "react-router-dom": "^5.2.0",
+    "styled-components": "^5.2.1"
   },
   "engines": {
     "node": ">=18.0.0 <=20.x.x",

